util/src/main/java/io/kubernetes/client/informer/cache/DeltaFIFO.java

```diff
@@ -0,0 +1,505 @@
+import static io.kubernetes.client.informer.cache.Deltas.DeltaType.Sync;
+
+import io.kubernetes.client.informer.cache.function.KeyFunc;
+import io.kubernetes.client.informer.cache.function.PopProcessFunc;
+import io.kubernetes.client.informer.exception.EmptyDeltasException;
+import io.kubernetes.client.util.common.Collections;
+import java.util.*;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// DeltaFIFO is a java portable of k/client-go's DeltaFIFO
+
+// NOTE(yue9944882): The DeltaFIFO inherits from a Object queue since that we are
+// actually reading from it as Deltas but writing it as Kubernetes object. this is
+// an issue existing in k/k.
+public class DeltaFIFO<ApiType> implements Queue<Object> {
+
+  private static final Logger log = LoggerFactory.getLogger(DeltaFIFO.class);
+
+  private KeyFunc keyFunc;
+
+  // the key of items would be KeyFunc value pls refer metaNamespaceKeyFunc
+  private Map<String, Deltas<ApiType>> items;
+
+  // an underlying queue storing incoming items' keys
+  private LinkedList<String> queue;
+
+  //
+  private Store<ApiType> knownObjects;
+
+  // populated is true if the first batch of items inserted by Replace() has
+  // been populated or Delete/Add/Update was called first.
+  private boolean populated = false;
+
+  // initialPopulationCount is the number of items inserted by the first call
+  // of Replace()
+  private int initialPopulationCount;
+
+  /** lock provides thread safety * */
+  private ReadWriteLock lock = new ReentrantReadWriteLock();
+
+  /** indicates if the store is empty * */
+  private Condition notEmpty;
+
+  /**
+   * Constructor.
+   *
+   * @param keyFunc the key func
+   * @param knownObjects the known objects
+   */
+  public DeltaFIFO(KeyFunc keyFunc, Store knownObjects) {
+    this.keyFunc = keyFunc;
+    this.knownObjects = knownObjects;
+    this.items = new HashMap<>();
+    this.queue = new LinkedList<>();
+    this.notEmpty = lock.writeLock().newCondition();
+  }
+
+  /**
+   * Add items to the delta FIFO.
+   *
+   * @param obj the obj
+   */
+  @Override
+  public void add(Object obj) {
+    lock.writeLock().lock();
+    try {
+      populated = true;
+      this.queueActionLocked(Deltas.DeltaType.Added, obj);
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Update items in the delta FIFO.
+   *
+   * @param obj the obj
+   */
+  @Override
+  public void update(Object obj) {
+    lock.writeLock().lock();
+    try {
+      populated = true;
+      this.queueActionLocked(Deltas.DeltaType.Updated, obj);
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Delete items from the delta FIFO.
+   *
+   * @param obj the obj
+   */
+  @Override
+  public void delete(Object obj) {
+    String id = this.keyOf(obj);
```

Comment:
shouldn't this be inside the lock?

Reply:
[https://github.com/kubernetes/kubernetes/blob/cdbc9d8567c177b8c6505181f755440792b9ff06/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L191-L197](https://github.com/kubernetes/kubernetes/blob/cdbc9d8567c177b8c6505181f755440792b9ff06/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L191-L197)

ref its original go code from k/k. it's fine to be out of the lock i suppose?

Reply:
this has no need to inside the lock, just a function call with the obj.